core/src/main/java/io/questdb/cairo/wal/WalPurgeJob.java

```diff
@@ -0,0 +1,403 @@
+        setWalPath(tableName, walId);
+        ff.iterateDir(path, this::deleteClosedSegmentsIter);
+    }
+
+    private void deleteOutstandingWalDirectories() {
+        for (PrimitiveIterator.OfInt it = discoveredWalIds.iterator(); it.hasNext(); ) {
+            walId = it.nextInt();
+            if (walsInUse.contains(walId)) {
+                deleteClosedSegments();
+            }
+            else {
+                deleteWalDirectory();
+            }
+        }
+    }
+
+    private boolean segmentIsReapable(int segmentId, int walsLatestSegmentId) {
+        return segmentId < walsLatestSegmentId;
+    }
+
+    private void deleteUnreachableSegmentsIter(long pUtf8NameZ, int type) {
+        if ((type == Files.DT_DIR) && matchesSegmentName(segmentName.of(pUtf8NameZ))) {
+            int segmentId = 0;
+            try {
+                segmentId = Numbers.parseInt(segmentName);
+            }
+            catch (NumericException _ne) {
+                return; // Ignore non-segment directory.
+            }
+            if (segmentIsReapable(segmentId, walsLatestSegmentId)) {
+                deleteSegmentDirectory(tableName, walId, segmentId);
+            }
+        }
+    }
+
+    private void deleteUnreachableSegments() {
+        for (int index = 0; index < walInfoDataFrame.size(); ++index) {
+            walId = walInfoDataFrame.walIds.get(index);
+            walsLatestSegmentId = walInfoDataFrame.segmentIds.get(index);
+            ff.iterateDir(setWalPath(tableName, walId), this::deleteUnreachableSegmentsIter);
+        }
+    }
+
+    private void accumDebugState() {
+        debugBuffer.clear();
+        debugBuffer.put("table=").put(tableName)
+                        .put(", discoveredWalIds=[");
+        for (PrimitiveIterator.OfInt it = discoveredWalIds.iterator(); it.hasNext(); ) {
+            final int walId = it.nextInt();
+            debugBuffer.put(walId);
+            if (walsInUse.contains(walId)) {
+                debugBuffer.put("(locked)");
+            }
+            if (it.hasNext()) {
+                debugBuffer.put(',');
+            }
+        }
+        debugBuffer.put("], walInfoDataFrame=[");
+        for (int index = 0; index < walInfoDataFrame.size(); ++index) {
+            debugBuffer.put('(').put(walInfoDataFrame.walIds.get(index)).put(',')
+                    .put(walInfoDataFrame.segmentIds.get(index)).put(')');
+            if (index < walInfoDataFrame.size() - 1) {
+                debugBuffer.put(',');
+            }
+        }
+        debugBuffer.put(']');
+    }
+
+    private boolean isWalTable(CharSequence tableName) {
+        final int tableNameLen = tableName.length();
+        final boolean result = ff.exists(setSeqTxnPath(tableName));
+        path.trimTo(tableNameLen).$();
+        return result;
+    }
+
+    private void broadSweepIter(long pUtf8NameZ, int type) {
+        if ((type == Files.DT_DIR) && isWalTable(tableName.of(pUtf8NameZ))) {
```

Comment:
Add try / catch at the method level so that failing of one table does not stop processing others.

Reply:
Done.